### ISSUE TITLE
RepoCloner and Flink deployment for license detection in Java.

### ISSUE THREAD

**Issue by michelescarlato** [fasten-project/fasten#426]
Hi Amir (@mir-am),

after a call with Sebastian(@proksch) we agreed upon running the current license detector (and feeder) for Java with the original requirements.

To provide Thomas (@tmortagne ) with license information in the FASTEN server we would need to spawn up the `RepoCloner` and `Flink` (JobManager and TaskManager). 

Thank you in advance,
Michele

**Comment by michelescarlato** [comment]
Hi Amir,

on Monster I cannot see them.
Where are they running?

**Comment by michelescarlato** [comment]
> Hi Michele,
> 
> Both `RepoCloner` and `Flink` have been running on the FASTEN clusters. 

Great!

> Your plug-in can listen to both `fasten.RepoCloner.out` and `fasten.RepoCloner.priority.out`. 
> Also, the cloned repositories are stored at `/mnt/fasten/data/repos/`.

Because of this cloned repository path change, the detector is not finding the repository cloned.
Could you please help me to troubleshoot it?

The license detector extracts the `repoPath` from the `kafkaRecord`, as you can see from this line:
https://github.com/fasten-project/fasten/blob/develop/analyzer/license-detector/src/main/java/eu/fasten/analyzer/licensedetector/LicenseDetectorPlugin.java#L72

I updated with the path you indicated (`/mnt/fasten/data/repos/`) the yamls configuration of the k8s StatefulSets ([detector mountPath](https://github.com/fasten-project/k8s-deployments/blob/license-detector/fasten/license-detector/base/detector/license-detector.yaml#L27), [detector hostPath](https://github.com/fasten-project/k8s-deployments/blob/license-detector/fasten/license-detector/base/detector/license-detector.yaml#L61), [feeder mountPath](https://github.com/fasten-project/k8s-deployments/blob/license-detector/fasten/license-detector/base/feeder/license-feeder.yaml#L28) and [feeder hostPath](https://github.com/fasten-project/k8s-deployments/blob/license-detector/fasten/license-detector/base/feeder/license-feeder.yaml#L69)  ).


From the following output, it seems pretty clear that the detector is still receiving the wrong `repoPath` in the Kafka topic coming from `fasten.SyncJava.out` (please keep in mind that we are using Flink to sync RepoCloner and MetadataDBJavaExtension plug-ins).

```
[2022-03-18 08:10:27,724] [INFO ] [main] [e.f.a.l.LicenseDetectorPlugin$LicenseDetector] - License detector started.
[2022-03-18 08:10:27,741] [WARN ] [main] [e.f.a.l.LicenseDetectorPlugin$LicenseDetector] - Path /mnt/fasten/repos/i/io.buoyant/k8s_2.11 does not denote a directory.
[2022-03-18 08:10:27,742] [INFO ] [main] [e.f.a.l.LicenseDetectorPlugin$LicenseDetector] - Retrieving outbound license from GitHub...
[2022-03-18 08:10:27,743] [WARN ] [main] [e.f.a.l.LicenseDetectorPlugin$LicenseDetector] - Path /mnt/fasten/repos/i/io.buoyant/k8s_2.11 does not denote a directory.
[2022-03-18 08:10:27,743] [WARN ] [main] [e.f.a.l.LicenseDetectorPlugin$LicenseDetector] - No outbound licenses were detected.
[2022-03-18 08:10:27,743] [INFO ] [main] [e.f.a.l.LicenseDetectorPlugin$LicenseDetector] - Scanning project in /mnt/fasten/repos/i/io.buoyant/k8s_2.11...
/usr/local/bin/scancode/src/cluecode/copyrights.py:3128: FutureWarning: Possible set difference at position 3
  re.MULTILINE | re.UNICODE
/usr/local/bin/scancode/lib/python3.7/site-packages/nltk/tag/sequential.py:532: FutureWarning: Possible nested set at position 7
  self._regexs = [(re.compile(regexp), tag,) for regexp, tag in regexps]
Usage: scancode [OPTIONS] <OUTPUT FORMAT OPTION(s)> <input>...
Try the 'scancode --help' option for help on options and arguments.

Error: Invalid value for "input": Path "/mnt/fasten/repos/i/io.buoyant/k8s_2.11" does not exist.
```


TBH, I have no idea where to set this in Flink. There are several parameters where paths are included.
Flink configs are [here](https://github.com/fasten-project/k8s-deployments/tree/license-detector/fasten/sync-job), and in particular I saw paths in:
- [jobamanager-deployment.yaml](https://github.com/fasten-project/k8s-deployments/blob/license-detector/fasten/sync-job/jobmanager-deployment.yaml);
- [taskamanager-deployment.yaml](https://github.com/fasten-project/k8s-deployments/blob/license-detector/fasten/sync-job/taskmanager-deployment.yaml);


Thanks in advance for any help you can provide here.

**Comment by mir-am** [comment]
By running the following command, it can be seen that `repoPath` has the correct path in `fasten.RepoCloner.out`.
```
kafkacat -b samos:9092/kafka -C -t fasten.RepoCloner.out -o beginning -e -q -c 1 | jq '.payload'
```
I think that you should run the Flink job again with another Kafa topic or ignore records that have wrong source code paths.

**Comment by mir-am** [comment]
@michelescarlato
Since I have not heard back after 5 days, I assume that the issue still persists.
To solve the issue, I need to stop both `jobmanager` and `taskmanager` deployments.
I will re-deploy them again with `fasten.SyncJava.new.out` as the output topic and it should contain the correct path to the repositories.

**Comment by mir-am** [comment]
I could not deploy `task-manager` again as it gives the following error:
```
The Deployment "flink-taskmanager-sync-job" is invalid: 
* spec.template.spec.containers[0].volumeMounts[0].name: Not found: "flink-final-config-volume"
* spec.template.spec.initContainers[0].volumeMounts[1].name: Not found: "flink-final-config-volume"
```
Maybe @amphioxis can provide help here as he recently modified the `taskmanager-deployment.yaml` [here](https://github.com/fasten-project/k8s-deployments/blob/license-detector/fasten/sync-job/taskmanager-deployment.yaml)

**Comment by amphioxis** [comment]
Hey @mir-am 
If you try to deploy `taskmanager` on the monster server, I just checked and I used the same config in Dec. which worked for me.
However, on our private cluster I had the following two lines after [line 59](https://github.com/fasten-project/k8s-deployments/blob/c6ab915115c822580abaf0628c10173effa7d870/fasten/sync-job/taskmanager-deployment.yaml#L59):
```
 - name: flink-final-config-volume
   emptyDir: { }
```
Maybe its worth a try...

But I am also not sure anymore if we changed the `taskmanager` on monster at all. If not, maybe try the original file.

**Comment by michelescarlato** [comment]
@mir-am Hi Amir, is there any specific reason why you are modifying the kafka topic produced?
As you can understand we will have to modify it accordingly.

I have also another question: "why when we deployed the detector and feeder in December, after spawning up the missing repo-cloner (indicated as a mandatory component in the detector and feeder pipeline diagram) we could see the license information in the FASTEN database, and now the repocloner path is generating this issue?

It has casually been changed ?"

Thanks, for your time answering and trying to help out with the issue generated by this different repository path,
M

**Comment by michelescarlato** [comment]
> If it is okay, I can delete the `fasten.SyncJava.out` topic and re-create it so that it contains only correct paths to the repositories. 


Wait a sec, did you fix this problem?

> The Deployment "flink-taskmanager-sync-job" is invalid: 
> * spec.template.spec.containers[0].volumeMounts[0].name: Not found: "flink-final-config-volume"
> * spec.template.spec.initContainers[0].volumeMounts[1].name: Not found: "flink-final-config-volume"


I thought still this was the main issue.

> Eariler this year, we have re-deployed the whole FASTEN pipeline and we decided to store pipeline-related data at `/mnt/fasten/data/`. Therefore, the previous repositories path do not exist anymore.

I wanted to be sure that really this breaks our license detection. Thanks for answering.

**Comment by proksch** [comment]
Why would this break the detection shouldn't this just be a changed config parameter somewhere? Or do you hardcode paths?

**Comment by michelescarlato** [comment]
The path are set in the k8s yamls file.

But as commented three days ago in the call with Giorgos @gdrosos , and as it is clearly shown at the beginning of this issue, changing the path in the license-detector and feeder has not been enough.

That's why we are talking about the Flink deployment.

Anyway, surely, not changing path the detection and feeding, as it was functioning in December, it would be still functioning.
Changing path, won't happen.

S. @proksch , why did we need to change this path?

**Comment by mir-am** [comment]
> > If it is okay, I can delete the `fasten.SyncJava.out` topic and re-create it so that it contains only correct paths to the repositories.
> 
> Wait a sec, did you fix this problem?
> 
> > The Deployment "flink-taskmanager-sync-job" is invalid:
> > 
> > * spec.template.spec.containers[0].volumeMounts[0].name: Not found: "flink-final-config-volume"
> > * spec.template.spec.initContainers[0].volumeMounts[1].name: Not found: "flink-final-config-volume"
> 
> I thought still this was the main issue.
> 
> > Eariler this year, we have re-deployed the whole FASTEN pipeline and we decided to store pipeline-related data at `/mnt/fasten/data/`. Therefore, the previous repositories path do not exist anymore.
> 
> I wanted to be sure that really this breaks our license detection. Thanks for answering.

Not yet. I need to fix the volume issues and then re-deploy Flink's job and task managers.
----------------
@proksch, Aboslulte paths to the repositories are stored in the topic `fasten.RepoCloner.(priority).out`.

**Comment by proksch** [comment]
@michelescarlato The original intent was to have a single place of derived data that we can delete whenever we reset the DB. We wanted to distinguish this from other types of data (most notably the Maven .m2 folder) that should be persisted and kept across DB resets.

**Comment by michelescarlato** [comment]
Hi @proksch and @mir-am , AFAIK @amphioxis will be out of the office (for exams) until 11/04. Martin reported the call with Amir, but still, I don't have very clear the situation of this Java pipeline.

We usually used a diagram similar to this one, to not lose track of the whole logic:
![image](https://user-images.githubusercontent.com/10910590/160521813-c4d81c69-c02b-44df-844a-4746da2dc23a.png)

@mir-am  please update this draw.io accordingly (we really need to be sure about the diagram architecture to be able to perform our intervention here). 

From my understanding the main points are two:
- the `repoPath` produced by the `RepoCloner` is changed (and probably Martin hardcoded the path in Flink, is that correct @mir-am ? ... that should be easily fixed inside of the Flink YAML k8s config files, from your side, Amir, right?)
- the new shape of the pipeline (modified after December 2021, when this license detection and feeding setup has been tested for Java) is introducing new topics to be consumed. It is that correct @mir-am (again, please provide us with an updated diagram, or it will be very difficult to understand correctly, I also noticed that the [Kafka topics](https://github.com/fasten-project/fasten/wiki/Kafka-Topics#fastenrepocloner) in the wiki are not updated), IMHO we should probably introduce two Flinks, having the following logic.

1. one instance that syncs the `fasten.RepoCloner.out` topic with `fasten.MetadataDBJavaExtension` (as shown [here](https://github.com/fasten-project/k8s-deployments/blob/license-detector/fasten/sync-job/jobmanager-deployment.yaml#L46-L49)).
2. and another one that syncs the fasten.RepoCloner.priority.out (@mir-am is that correct?) topic with `fasten.MetadataDBJavaExtension.out` (@mir-am , is this correct? ... or is it introduced the keyword `priority` somewhere?), producing a kind of `fasten.SyncJava.priority.out`.
3. After introducing the second Flink deployment we will also need to deploy a second instance of the `java-license-detector` that consumes the `fasten.SyncJava.priority.out`, and produces a `fasten.LicenseDetector.priority.out`.
4. After introducing the second instance of the `java-license-detector`, we should introduce a second instance of the `java-license-feeder` which consumes the fasten.LicenseDetector.priority.out produced by the `java-license-detector` .

Does this make sense to you @mir-am ?

Can you please fix Flink for us (since we will not have Martin available for the next 13 days), possibly adding this second instance of Flink used for priorities, and I will take care of introducing the priorities instances of the detector and feeder?

Also, please, can you update accordingly the diagram (here there is the [draw.io](https://drive.google.com/file/d/1hrYALK7kZTRIBEnfXAgK-aeNeZ5ZREVy/view?usp=sharing) file, I just shared it with your TUDelft account) ?

Thanks for your understanding and collaboration,
M

**Comment by mir-am** [comment]
Hi Michele,

- Note that `repoPath` is determined by `RepoCloner`, not the Flink job. Also, the `repoPath` now points to `/mnt/fasten/data/repos`.
- As you mentioned, you can have two Flnk jobs, one of which syncs the normal topics, and the other one syncs the priority topics. Simply, you will have a flow like:
(`fasten.RepoCloner.out` + `fasten.MetadataDBJavaExtension.out`) -> `fasten.SyncJava.out`
(`fasten.RepoCloner.priority.out` + `fasten.MetadataDBJavaExtension.priority.out`) -> `fasten.SyncJava.priority.out`

- Regarding the data flow diagram, you can just add one more arrow next to every depicted arrow and add `(.priority)` to the names of the Kafka topics, e.g., `fasten.OPAL(.priority).out`.

-----------

That being said, I will try to deploy two instances of the Flink jobs as illustrated above.

**Comment by michelescarlato** [comment]
Hi @mir-am , I am going to close this because I saw some messages consumed correctly by the license-detector (I stored the logs on Monster and if you indicated to me the right place to push them, we can have a look later to be sure that we are experiencing the expected behavior).

We are instead experiencing an issue with the feeder, that IMHO can be linked to the connection with the Java DB.
The issue is [42](https://github.com/fasten-project/k8s-deployments/issues/42).
